server/src/main/java/org/elasticsearch/repositories/blobstore/ShardSnapshotWorkers.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.repositories.blobstore;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionRunnable;
+import org.elasticsearch.common.CheckedBiConsumer;
+import org.elasticsearch.common.UUIDs;
+import org.elasticsearch.repositories.SnapshotShardContext;
+
+import java.io.IOException;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.Executor;
+import java.util.concurrent.PriorityBlockingQueue;
+import java.util.function.Consumer;
+
+import static org.elasticsearch.index.snapshots.blobstore.BlobStoreIndexShardSnapshot.FileInfo;
+
+/**
+ * ShardSnapshotWorkers performs snapshotting tasks in the order dictated by the PriorityQueue of snapshot tasks.
+ * Each enqueued shard to snapshot results in one @{@link ShardSnapshotTask} and zero or more @{@link FileSnapshotTask}.
+ */
+public class ShardSnapshotWorkers {
+    private static final Logger logger = LogManager.getLogger(ShardSnapshotWorkers.class);
+
+    private final int maxWorkers;
+    private final Object mutex = new Object();
+    private final BlockingQueue<ShardSnapshotTask> snapshotTasks = new PriorityBlockingQueue<>();
+    private final Executor executor;
+    private final Consumer<SnapshotShardContext> shardSnapshotter;
+    private final CheckedBiConsumer<SnapshotShardContext, FileInfo, IOException> fileSnapshotter;
+    private volatile int workerCount = 0;
+
+    class ShardSnapshotTask implements Comparable<ShardSnapshotTask> {
+        protected final SnapshotShardContext context;
+
+        ShardSnapshotTask(SnapshotShardContext context) {
+            this.context = context;
+        }
+
+        public SnapshotShardContext context() {
+            return context;
+        }
+
+        public void run() {
+            shardSnapshotter.accept(context);
+        }
+
+        public short priority() {
+            return 1;
+        }
+
+        @Override
+        public int compareTo(ShardSnapshotTask other) {
```

Comment:
I think we should mark this method `final` since overriding is likely bug-prone.

Reply:
Done!